src/sentry/buffer/redis.py

```diff
@@ -208,6 +208,56 @@ def get(
             col: (int(results[i]) if results[i] is not None else 0) for i, col in enumerate(columns)
         }
 
+    def get_redis_connection(self, key: str) -> Any | None:
+        # TODO type RedisPipeline instead of using Any here
+        if is_instance_redis_cluster(self.cluster, self.is_redis_cluster):
+            conn = self.cluster
+        elif is_instance_rb_cluster(self.cluster, self.is_redis_cluster):
+            conn = self.cluster.get_local_client_for_key(key)
+        else:
+            raise AssertionError("unreachable")
+
+        pipe = conn.pipeline()
+        return pipe
+
+    def push_to_set(self, key: str, value: list[int, int] | int) -> None:
+        pending_key = self._make_pending_key_from_key(key)
+        pipe = self.get_redis_connection(pending_key)
+        pipe.sadd(key, value)
+        pipe.expire(key, self.key_expire)
+        pipe.execute()
+
+    def get_set(self, key: str) -> None:
+        pending_key = self._make_pending_key_from_key(key)
+        pipe = self.get_redis_connection(pending_key)
+        pipe.smembers(key)
+        return pipe.execute()
+
+    def push_to_hash(
+        self, key: str, field: dict[str, int], value: dict[str, dict[int, int]]
+    ) -> None:
+        pending_key = self._make_pending_key_from_key(key)
+        pipe = self.get_redis_connection(pending_key)
+        pipe.hsetnx(key, str(field["project_id"]), json.dumps(value))
```

Comment:
This doesn't support adding more data like if a rule fires again in the same project. I think we'll have to change the data structure because right now it checks if the key already exists (which is the project id) and if so, doesn't add it. 

It's a bit confusing because it comes back as a list  (`[{'1': '{"2:3":4}'}]`) so I'd hope I could add to the list, but I'm not entering the data as a list, but a (`json.dumps`ed) dictionary 🤔

Reply:
Also having to do `field["project_id"]` will need to change, but I need `field` to be a dictionary for now otherwise creating the pending key fails.